src/herder/HerderImpl.cpp

```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
     }
 }
 
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+void
+HerderImpl::maybeHandleUpgrade()
+{
```

Comment:
mmm, `getSorobanNetworkConfig` doesn't fail if protocolVersion < 20, but instead creates an empty config, right? 
https://github.com/stellar/stellar-core/blob/0eab126b4343cf7d93af2cc9ca97c5bf7728ee4a/src/ledger/LedgerManagerImpl.cpp#L1148

So I believe this function should be a no-op if we're running protocol 19 (as confirmed by "upgrade to generalized tx set in network" test which runs the network at proto 19 for some time, then upgrades to 20). That being said, it does seem error-prone (e.g. what if this logic handles default values incorrectly), so I agree we can add a protocol check just in case.

Reply:
Yeah it's probably fine as is, I just get a little spooked calling anything with Soroban in the name pre protocol 20.

Reply:
yeah, I also opened https://github.com/stellar/stellar-core/issues/3828 to prohibit retrieving soroban config prior to protocol 20